libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -239,64 +272,164 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
 
-	ostringstream out;
+	if (FunctionDefinition const* constructor = _contract.constructor())
+		for (auto const& modifier: constructor->modifiers())
+		{
+			FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+					modifier->name()->annotation().referencedDeclaration
+			)->constructor();
 
-	FunctionDefinition const* constructor = _contract.constructor();
-	if (constructor && !constructor->isPayable())
-		out << callValueCheck();
+			if (baseConstructor && modifier->arguments() && !modifier->arguments()->empty())
+				m_baseConstructorParameters[baseConstructor] = listParamNames(baseConstructor, modifier->arguments());
+		}
 
-	for (ContractDefinition const* contract: reverse(_contract.annotation().linearizedBaseContracts))
+	for (ASTPointer<InheritanceSpecifier> const& base: _contract.baseContracts())
 	{
-		out <<
-			"\n// Begin state variable initialization for contract \"" <<
-			contract->name() <<
-			"\" (" <<
-			contract->stateVariables().size() <<
-			" variables)\n";
-
-		IRGeneratorForStatements generator{m_context, m_utils};
-		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!variable->isConstant() && !variable->immutable())
-				generator.initializeStateVar(*variable);
-		out << generator.code();
-
-		out << "// End state variable initialization for contract \"" << contract->name() << "\".\n";
+		FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+				base->name().annotation().referencedDeclaration
+		)->constructor();
+
+		if (baseConstructor && base->arguments() && !base->arguments()->empty())
+			m_baseConstructorParameters[baseConstructor] = listParamNames(baseConstructor, base->arguments());
 	}
 
-	if (constructor)
-	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+	return generator.code();
+}
+
+string IRGenerator::initStateVariables(ContractDefinition const& _contract)
+{
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (VariableDeclaration const* variable: _contract.stateVariables())
+		if (!variable->isConstant() && !variable->immutable())
+			generator.initializeStateVar(*variable);
 
-		Whiskers t(R"X(
-			let programSize := datasize("<object>")
-			let argSize := sub(codesize(), programSize)
+	return generator.code();
+}
+
+string IRGenerator::copyParametersToMemory(ContractDefinition const& _contract)
+{
+	ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
 
-			let memoryDataOffset := <allocate>(argSize)
-			codecopy(memoryDataOffset, programSize, argSize)
+	string params = suffixedVariableNameList(
+		"param_" + m_context.functionName(*_contract.constructor()) + "_",
+		0,
+		_contract.constructor()->parameters().size()
+	);
+
+	Whiskers t(R"X(
+		let programSize := datasize("<object>")
```

Comment:
If possible, yul code should always be in functions, because then we will not get into trouble with overlapping variable scopes and having to remember the names of variables that we will use later on.

I think this snippet could be turned into a function that returns `param_i` instead.

Reply:
This part was moved to util functions and it is separate functions that returns parameters now